source/extensions/filters/http/cache/http_cache_utils.cc

```diff
@@ -0,0 +1,188 @@
+  // qdtext          = HTAB / SP /%x21 / %x23-5B / %x5D-7E / obs-text
+  // obs-text        = %x80-FF
+  // quoted-pair     = "\" ( HTAB / SP / VCHAR / obs-text )
+  // VCHAR           =  %x21-7E  ; visible (printing) characters
+  SystemTime::duration max_age = SystemTime::duration::zero();
+  bool found_s_maxage = false;
+  while (!cache_control.empty()) {
+    // Each time through the loop, we eat one cache-directive. Each branch
+    // either returns or completely eats a cache-directive.
+    if (ConsumePrefix(&cache_control, "no-cache")) {
+      if (eatToken(cache_control)) {
+        // The token wasn't no-cache; it just started that way, so we must
+        // finish eating this cache-directive.
+        if (ConsumePrefix(&cache_control, "=")) {
+          eatDirectiveArgument(cache_control);
+        }
+      } else {
+        // Found a no-cache directive, so validation is required.
+        return SystemTime::duration::zero();
+      }
+    } else if (ConsumePrefix(&cache_control, "s-maxage=")) {
+      max_age = eatLeadingDuration(cache_control);
+      found_s_maxage = true;
+      cache_control = StripLeadingAsciiWhitespace(cache_control);
+      if (!cache_control.empty() && cache_control[0] != ',') {
+        // Unexpected text at end of directive
+        return SystemTime::duration::zero();
+      }
+    } else if (!found_s_maxage && ConsumePrefix(&cache_control, "max-age=")) {
+      max_age = eatLeadingDuration(cache_control);
+      if (!cache_control.empty() && cache_control[0] != ',') {
+        // Unexpected text at end of directive
+        return SystemTime::duration::zero();
+      }
+    } else if (eatToken(cache_control)) {
+      // Unknown directive--ignore.
+      if (ConsumePrefix(&cache_control, "=")) {
+        eatDirectiveArgument(cache_control);
+      }
+    } else {
+      // This directive starts with illegal characters. Require validation.
+      return SystemTime::duration::zero();
+    }
+    // Whichever branch we took should have consumed the entire cache-directive,
+    // so we just need to eat the delimiter and optional whitespace.
+    ConsumePrefix(&cache_control, ",");
+    cache_control = StripLeadingAsciiWhitespace(cache_control);
+  }
+  return max_age;
+}
+
+SystemTime Utils::httpTime(const Http::HeaderEntry* header_entry) {
+  if (!header_entry) {
+    return {};
+  }
+  absl::Time time;
+  const std::string input(header_entry->value().getStringView());
+
+  // Acceptable Date/Time Formats per
+  // https://tools.ietf.org/html/rfc7231#section-7.1.1.1
+  //
+  // Sun, 06 Nov 1994 08:49:37 GMT    ; IMF-fixdate
+  // Sunday, 06-Nov-94 08:49:37 GMT   ; obsolete RFC 850 format
+  // Sun Nov  6 08:49:37 1994         ; ANSI C's asctime() format
+  const std::array<std::string, 3> rfc7231_date_formats = {
+      "%a, %d %b %Y %H:%M:%S GMT", "%A, %d-%b-%y %H:%M:%S GMT", "%a %b %e %H:%M:%S %Y"};
+  for (const std::string& format : rfc7231_date_formats) {
```

Comment:
it's very annoying that absl::ParseTime takes a std::string as the format arg rather than an absl::string_view.

So it looks every time httpTime() is called, we'll have to construct 3 format strings.

Can you add a TODO, e.g. lazy-init these strings once rather than stack-allocate them?

Reply:
done